File: lib/leetcode/medium/product_of_array_except_self.dart
List<int> productExceptSelf(List<int> nums) {
  final n = nums.length;
  List<int> answers = List.filled(n, 1);

  // Lets find the prefix value;
  int prefix = 1;
  for (int i = 0; i < n; i++) {
    answers[i] *= prefix;
    prefix *= nums[i];
  }

  // Lefts find the postfix value;
  int postfix = 1;
  for (int i = n - 1; i >= 0; i--) {
    answers[i] *= postfix;
    postfix *= nums[i];
  }

  return answers;
}

void main() {
  List<int> nums1 = [1, 2, 3, 4];
  print(
      'Output for nums1: ${productExceptSelf(nums1)}'); // Output: [24, 12, 8, 6]

  List<int> nums2 = [-1, 1, 0, -3, 3];
  print(
      'Output for nums2: ${productExceptSelf(nums2)}'); // Output: [0, 0, 9, 0, 0]
}

File: lib/leetcode/medium/ProductOfArrayExceptSelf.java
package lib.leetcode.medium;

import java.util.Arrays;

public class ProductOfArrayExceptSelf {

    public int[] productExceptSelf(int[] nums) {
        final  int n = nums.length;
         int[] answers = new int[n];
         Arrays.fill( answers, 1);

        // lets find all prefix value first
        int prefix = 1;
        for (int i = 0; i<n; i++) {
            answers[i] *= prefix;
            prefix *= nums[i];
        }

        // Lets find all the postfix values with prefix value already found
        int postfix = 1;
        for(int i = n -1; i >= 0; i--) {
            answers[i] *= postfix;
            postfix *= nums[i];
        }

        return answers;
    }
    
}
